vncviewer/GestureHandler.h

```diff
@@ -0,0 +1,74 @@
+/* Copyright 2019 Aaron Sowry for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifndef __GESTUREHANDLER_H__
+#define __GESTUREHANDLER_H__
+
+#include <vector>
+
+#include <X11/Xlib.h>
+#include <X11/extensions/XInput2.h>
+
+// Internal state bitmasks
+#define GH_LEFTBTN     1
+#define GH_RIGHTBTN    2
+#define GH_MIDDLEBTN   4
+#define GH_SCROLL      8
+#define GH_ZOOM        16
+#define GH_NOGESTURE   128
+#define GH_UNDEFINED   (32 | 64)
+
+#define GH_INITSTATE   (255 & ~GH_UNDEFINED)
+
+// Timeout when waiting for gesture
+#define GH_TOUCHDELAY  0.25 // s
+
+enum GHEventType {
+  GH_ButtonPress,
+  GH_ButtonRelease,
+  GH_TouchUpdate,
```

Comment:
Did you look at having a more gesture based abstraction here? I.e. events like `GH_SingleTap`, `GH_Drag`, H_Pinch`? I think it might make it clearer what it is the logic is trying to detect. Then we can let a different layer define what that means in terms of pointer events.

Reply:
I'm still thinking about this. I'm leaning towards just having a `GH_GestureBegin` event, with a member which describes which gesture has begun. The corresponding `GH_GestureEnd` wouldn't need this information since only one gesture is active at a time. Note that things like a single tap or click are essentially a gestures now since they're dependent on a timeout.

Reply:
That sounds like a very nice approach.